### ISSUE TITLE
Type of __doc__ should not be `str | None` when docstring exists

### ISSUE THREAD

**Issue by millerdev** [astral-sh/ty#3877]
### Summary

ty reports

> Attribute `splitlines` is not defined on `None` in union `str | None` ty([unresolved-attribute](https://ty.dev/rules#unresolved-attribute))

when `__doc__` is clearly not `None`. This creates a broken window (error in module) when there is no error.

<img width="885" height="297" alt="Image of error in VS Codium" src="https://github.com/user-attachments/assets/68fd9afe-f3db-4677-b190-6c2ccbbd2eb9" />

To clean up the error one could make this working code ugly with `# ty:ignore[unresolved-attribute]` or resort to unnecessary logic like `(__doc__ or "").splitlines()`. It should be possible to manipulate a defined `__doc__` without resorting to defensive typing measures.

Reproduction code:
```py
"""Am I here at all?

The type checker cannot say.
And yet here I am.
"""
print(__doc__.splitlines()[0])
```

Similar typed code does not report an error:
```py
value: str | None = "No error here."
value.splitlines()
```

### Version

ty 0.0.54

**Comment by sharkdp** [comment]
Thank you for reporting this. The type of `__doc__` currently simply comes from typeshed: [`ModuleType.__doc__`](https://github.com/python/typeshed/blob/b9090e99745ac1511d8efd828622b11a9a3623e8/stdlib/types.pyi#L338).

I guess it's reasonable to try and detect cases where `__doc__` is definitely not `None`.

For comparison: pyright also errors on the snippet, pyrefly and mypy do not. Pyrefly seems to be the only type checker that actually detects the presence of the module-level docstring.

**Comment by millerdev** [comment]
> For comparison: pyright also errors on the snippet

My understanding is that Pylance uses pyright internally, but Pylance does not report an error. Not sure if that is a helpful detail.

**Comment by Fogapod** [comment]
```sh
$ python a.py
Am I here at all?

$ python -OO a.py
Traceback (most recent call last):
  File "/home/u/a.py", line 6, in <module>
    print(__doc__.splitlines()[0])
          ^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'splitlines'
```

**Comment by AlexWaygood** [comment]
yes, true, in `-OO` mode Python strips all docstrings at runtime as a memory optimisation. So if you want your code to be type-safe in `-OO` mode, you'd never want the type checker to infer the type of `__doc__` as `str` just because the type checker sees a docstring in the source code.

As @sharkdp showed, however, most other type checkers do not follow that logic. I'm not totally sure how common it is to run Python in `-OO` mode.

**Comment by Fogapod** [comment]
could it be a configuration option maybe? target optimization level, assume `-OO` by default. opt-in for non-optimized where `__doc__` can be `str`

Edit: nevermind, -OO by default would mean `Literal[None]` always

**Comment by sharkdp** [comment]
> could it be a configuration option maybe?

In [this comment](https://github.com/astral-sh/ruff/pull/26505#discussion_r3512441582), I proposed that we add the following justification for adding this feature:

*Docstrings are stripped in `-OO` optimized mode, but [...] we assume that the existence of an actual docstring AND the usage of `__doc__` is reason enough to believe that [the attribute] will exist at runtime.*

On the other hand, I understand that keeping `str | None` could help someone find an actual problem before deploying their app to production. It would be great if we could avoid introducing a configuration setting for this specific thing. Maybe there is a more general setting that would make sense?

**Comment by carljm** [comment]
I think running with `-OO` is quite rare in practice, and it's better for us to use the rationale @sharkdp gives.

If you know that you are running your app with `-OO` in prod, then the rule you need is quite simple and can be implemented as an opt-in rule in Ruff, it doesn't need type information: you should never use `__doc__` at all.
